### PR TITLE
SIP-138 Update Volume Tracking Event

### PR DESCRIPTION
**Pull request by dvd-schwrtz**
https://sips.synthetix.io/sips/sip-138

### CODE REVIEW

**Review by jacko125 — APPROVED** [review]

File: contracts/Exchanger.sol [hunk 1]
```diff
@@ -405,7 +406,7 @@ contract Exchanger is Owned, MixinSystemSettings, IExchanger {
 
         _processTradingRewards(fee, originator);
 
-        _emitTrackingEvent(trackingCode, destinationCurrencyKey, amountReceived);
+        _emitTrackingEvent(trackingCode, destinationCurrencyKey, amountReceived, fee);
```
Comment (jacko125) [inline contracts/Exchanger.sol:409]: @dvd-schwrtz Just note that this fee is the nominal sUSD fee amount of the trade, not the fee % rate.

**Review by jacko125 — APPROVED** [review]
lgtm

**Review by i-stam — APPROVED** [review]

**Review by ajsantander — CHANGES_REQUESTED** [review]

File: contracts/Exchanger.sol [hunk 2]
```diff
@@ -32,6 +32,7 @@ interface ISynthetixInternal {
         bytes32 trackingCode,
         bytes32 toCurrencyKey,
         uint256 toAmount
+        uint fee
```
[the remark this answers is not in the document]
Comment (ajsantander) [reply contracts/Exchanger.sol:35]: Same for BaseSynthetix

File: test/contracts/BaseSynthetix.js [hunk 3]
```diff
@@ -287,6 +291,7 @@ contract('BaseSynthetix @ovm-skip', async accounts => {
 						trackingCode: trackingCode,
 						toCurrencyKey: currencyKey1,
 						toAmount: amount1,
+						fee: amount2,
```
Comment (ajsantander) [inline test/contracts/BaseSynthetix.js:294]: We should probably add tests for this event in Exchanger.spec.js and make sure that it emits with the expected values during an exchange.

File: contracts/BaseSynthetix.sol [hunk 4]
```diff
@@ -401,9 +401,10 @@ contract BaseSynthetix is IERC20, ExternStateToken, MixinResolver, ISynthetix {
     function emitExchangeTracking(
         bytes32 trackingCode,
         bytes32 toCurrencyKey,
-        uint256 toAmount
+        uint256 toAmount,
+        uint fee
```
[the remark this answers is not in the document]
Comment (ajsantander) [reply contracts/BaseSynthetix.sol:405]: The EXCHANGE_TRACKING_SIG declaration underneath the event declaration also needs to be updated to:
`ExchangeTracking(bytes32,bytes32,uint256,uint256)`

File: contracts/Exchanger.sol [hunk 5]
```diff
@@ -429,16 +430,17 @@ contract Exchanger is Owned, MixinSystemSettings, IExchanger {
...
+            _emitTrackingEvent(trackingCode, destinationCurrencyKey, amountReceived, fee);
         }
     }
 
     function _emitTrackingEvent(
         bytes32 trackingCode,
         bytes32 toCurrencyKey,
-        uint256 toAmount
+        uint256 toAmount,
+        uint fee
```
Comment (ajsantander) [inline contracts/Exchanger.sol:441]: uint256

File: contracts/Exchanger.sol (hunk 1 above)
Comment (dvd-schwrtz) [reply contracts/Exchanger.sol:409]: @jacko125 The notes in the _exchange method say that it is based in the destination currency key amount

        // Note: `fee` is denominated in the destinationCurrencyKey.
        (amountReceived, fee, exchangeFeeRate, sourceRate, destinationRate) = _getAmountsForExchangeMinusFees(
            sourceAmountAfterSettlement,
            sourceCurrencyKey,
            destinationCurrencyKey
        );

Comment (justinjmoses) [reply contracts/Exchanger.sol:409]: wrong section - see `function _exchange`

**Review by ajsantander — APPROVED** [review]
Looks good!

**Review by i-stam — APPROVED** [review]

File: test/contracts/Exchanger.spec.js [hunk 6]
```diff
@@ -1825,12 +1827,12 @@ contract('Exchanger (spec tests)', async accounts => {
 							toAddress: account1,
 						});
 
-						const trackingEvent = txn.logs.find(log => log.event === 'ExchangeTracking');
-						assert.eventEqual(trackingEvent, 'ExchangeTracking', {
-							trackingCode,
-							toCurrencyKey: toBytes32('sAUD'),
-							toAmount: sAUDBalance,
-						});
```
Comment (justinjmoses) [inline test/contracts/Exchanger.spec.js:1833]: be careful here - you broke the parsing logic by removing this.

File: test/contracts/Exchanger.spec.js [hunk 7]
```diff
@@ -2175,13 +2177,13 @@ contract('Exchanger (spec tests)', async accounts => {
 								const usdFeeAmount = await exchangeRates.effectiveValue(sAUD, fee, sUSD);
 								assert.bnEqual(usdFeeAmount, feePeriodZero.feesToDistribute);
 
-								// Assert the tracking event is fired.
-								const trackingEvent = txn.logs.find(log => log.event === 'ExchangeTracking');
-								assert.eventEqual(trackingEvent, 'ExchangeTracking', {
-									trackingCode,
-									toCurrencyKey: toBytes32('sAUD'),
-									toAmount: sAUDBalance,
-								});
```
Comment (justinjmoses) [inline test/contracts/Exchanger.spec.js:2184]: and here as well.

File: test/contracts/Exchanger.spec.js (hunk 6 above)
Comment (justinjmoses) [reply test/contracts/Exchanger.spec.js:1833]: you can tell because the formatter on L1816 didn't fix the indents for you on save.

**Pull request merged**